Opacity and root element background image doesn't render.

mattwoodrow opened the pull request:
#### 80a5258521a832779b46bdd6bb04368f36882e5c
<pre>
Opacity and root element background image doesn&apos;t render.
<a href="https://bugs.webkit.org/show_bug.cgi?id=261486">https://bugs.webkit.org/show_bug.cgi?id=261486</a>
&lt;rdar://115396444&gt;

Reviewed by Simon Fraser.

There are two bugs fixed here:

One is that transparencyClipBox doesn&apos;t account for the background rendering area of the root element when computing the clip
to push for the opacity group.

The other is that repaintLayerRectsForImage always issued a repaint request to the RenderView if the background image on the root
element changed, which is incorrect if the root element is composited.

The remaining test failure is because we don&apos;t apply opacity when we have background-attachment:fixed, and the root element is
composited. Bug 245032 should fix that in the near future.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-004-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-004.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/support/transform-triangle-left.svg: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::repaintLayerRectsForImage):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::transparencyClipBox):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: <a href="https://commits.webkit.org/268156@main">https://commits.webkit.org/268156@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e25cdbf4420bf846d388001238c894d3eef089e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18867 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19210 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19815 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20736 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17654 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19059 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22513 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19345 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19092 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/19223 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16435 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21620 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16438 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17199 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23624 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17472 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/17370 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21547 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/17957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17026 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17032 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4481 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21391 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/17793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -2079,6 +2079,15 @@ bool RenderBox::repaintLayerRectsForImage(WrappedImagePtr image, const FillLayer
                         std::max(layerRenderer->width() + layerRenderer->horizontalMarginExtent() + layerRenderer->borderLeft() + layerRenderer->borderRight(), rw),
                         std::max(layerRenderer->height() + layerRenderer->verticalMarginExtent() + layerRenderer->borderTop() + layerRenderer->borderBottom(), rh));
+
+                    // If we're drawing the root background, then we
```
Oddly wrapped comment; could wrap wider.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2284,8 +2284,10 @@ static LayoutRect transparencyClipBox(const RenderLayer& layer, const RenderLaye
+
+    OptionSet<RenderLayer::CalculateLayerBoundsFlag> flags = transparencyBehavior == HitTestingTransparencyClipBox ? RenderLayer::UseFragmentBoxesIncludingCompositing : RenderLayer::UseFragmentBoxesExcludingCompositing;
+    flags.add(RenderLayer::IncludeRootBackgroundPaintingArea);
+    LayoutRect clipRect = layer.boundingBox(rootLayer, layer.offsetFromAncestor(rootLayer), flags);
```
auto

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4844,6 +4846,15 @@ LayoutRect RenderLayer::localBoundingBox(OptionSet<CalculateLayerBoundsFlag> fla
         } else
             result = box->visualOverflowRect();
+
+        if (flags & IncludeRootBackgroundPaintingArea && renderer().isDocumentElementRenderer()) {
```
`flags.contains()`

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4844,6 +4846,15 @@ LayoutRect RenderLayer::localBoundingBox(OptionSet<CalculateLayerBoundsFlag> fla
+            // If the root layer becomes composited (e.g. because some descendant with negative z-index is composited),
+            // then it has to be big enough to cover the viewport in order to display the background. This is akin
+            // to the code in RenderBox::paintRootBoxFillLayers().
+            const LocalFrameView& frameView = renderer().view().frameView();
```
auto

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4844,6 +4846,15 @@ LayoutRect RenderLayer::localBoundingBox(OptionSet<CalculateLayerBoundsFlag> fla
+            // to the code in RenderBox::paintRootBoxFillLayers().
+            const LocalFrameView& frameView = renderer().view().frameView();
+            result.setWidth(std::max(result.width(), frameView.contentsWidth() - result.x()));
+            result.setHeight(std::max(result.height(), frameView.contentsHeight() - result.y()));
```
Do we ever end up making a transparency layer with this size? Could it become huge, or will we still clip to tile boundaries?

mattwoodrow replied:
The transparency layer is restricted to the clip rect on the context, which includes the bounds of the context.

mattwoodrow committed:
Opacity and root element background image doesn't render.
https://bugs.webkit.org/show_bug.cgi?id=261486
<rdar://115396444>

Reviewed by Simon Fraser.

There are two bugs fixed here:

One is that transparencyClipBox doesn't account for the background rendering area of the root element when computing the clip
to push for the opacity group.

The other is that repaintLayerRectsForImage always issued a repaint request to the RenderView if the background image on the root
element changed, which is incorrect if the root element is composited.

The remaining test failure is because we don't apply opacity when we have background-attachment:fixed, and the root element is
composited. Bug 245032 should fix that in the near future.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-004-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/root-element-background-image-transparency-004.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/compositing/support/transform-triangle-left.svg: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::repaintLayerRectsForImage):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::transparencyClipBox):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: https://commits.webkit.org/268156@main
